src/devices/Nmea0183/Ais/AisParser.cs

```diff
@@ -0,0 +1,284 @@
+                fullData = fullData.Remove(0, thisMessageLength);
+                parts.Add(thisMessagePayLoad);
+                // Only the last block of a multi-part message needs padding
+                if (blockNumber == numBlocks)
+                {
+                    parts.Add(paddedBits.ToString(CultureInfo.InvariantCulture));
+                }
+                else
+                {
+                    parts.Add("0");
+                }
+
+                RawSentence rs = new RawSentence(TalkerId.Ais, GeneratedSentencesId, parts, DateTimeOffset.UtcNow);
+                ret.Add(rs);
+                blockNumber++;
+            }
+
+            return ret;
+        }
+
+        private bool IsValidAisSentence(RawSentence rs)
+        {
+            return rs.Valid && rs.Fields.Length == 6 && (rs.SentenceId == VdoId || rs.SentenceId == VdmId) &&
+                   rs.TalkerId == TalkerId.Ais;
+        }
+
+        private Payload? DecodePayload(string encodedPayload, int numFragments, int fragmentNumber, int messageId, int numFillBits)
+        {
+            if (numFragments == 1)
+            {
+                var decoded = _payloadDecoder.Decode(encodedPayload, numFillBits);
+                return decoded;
+            }
+
+            lock (_fragments)
+            {
+                if (fragmentNumber == 1)
+                {
+                    // Note this clears any previous message parts, which is intended (apparently the previous group with this messageId was never completed)
+                    var l = new List<string>(numFragments) { encodedPayload };
+                    _fragments[messageId] = l;
+                    return null;
+                }
+
+                if (fragmentNumber <= numFragments)
+                {
+                    if (_fragments.TryGetValue(messageId, out var existingParts) && existingParts.Count == fragmentNumber - 1)
+                    {
+                        existingParts.Add(encodedPayload);
+                    }
+                    else
+                    {
+                        // Message is incomplete or out of order -> drop it
+                        _fragments.Remove(messageId);
+                        return null;
+                    }
+                }
+
+                if (fragmentNumber == numFragments)
+                {
+                    if (_fragments.TryGetValue(messageId, out var existingParts) &&
+                        existingParts.Count == numFragments)
+                    {
+                        // The collection is complete.
+                        encodedPayload = string.Join(string.Empty, existingParts);
+                        _fragments.Remove(messageId);
+                        return _payloadDecoder.Decode(encodedPayload, numFillBits);
+                    }
+                }
+
+                return null; // More parts expected
+            }
+        }
+
+        public int ExtractChecksum(string sentence, int checksumIndex)
```

Comment:
nit: checksumIndex seems to be misnamed given we're adding `+`, probably should be called `dataLength` or something

Reply:
No, that's right here, it's the index of the checksum in the input string, the +1 is for the checksum delimiter.